### ISSUE TITLE
Strange behavior with files containing `__` 

### ISSUE THREAD

**Issue by indigoviolet** [aai-institute/lakefs-spec#297]
### What happened?

```
❯ lakectl fs ls -r lakefs://sample/test-venky/data/foo/
object          2024-12-12 23:32:03 -0800 PST    2.7 kB          data/foo/bar/axe.just
object          2024-12-12 23:26:00 -0800 PST    0 B             data/foo/bar/quux/widg.txt
object          2024-12-12 23:26:00 -0800 PST    0 B             data/foo/bar__baz.txt
```

```
p = UPath("lakefs://sample/test-venky/data/foo", use_listings_cache=False)
list(p.fs.find(p, withdirs=True))
---
['sample/test-venky/data/foo',
 'sample/test-venky/data/foo/bar/',
 'sample/test-venky/data/foo/bar__baz.txt']
```
 
Where did `foo/bar/*/` go? 

`p.fs.walk()` has the same behavior, but `p.fs.ls(p, recursive=True)` returns the right information

The files do exist if you drill down and look explicitly for them

```
p = UPath("lakefs://sample/test-venky/data/foo/bar/axe", use_listings_cache=False)
p.fs.ls(p)
---
[{'checksum': 'a87e010a318457fd39f9525e6d1c576d',
  'content-type': 'application/octet-stream',
  'mtime': 1734075123,
  'name': 'sample/test-venky/data/foo/bar/axe.just',
  'size': 2696,
  'type': 'object'}]
```

A further problem is that this leads to fs.walk reporting non-existent files:

```
p = UPath("lakefs://sample/test-venky/data/foo/", use_listings_cache=False)
list(p.fs.walk(p))
---
[('sample/test-venky/data/foo', ['bar'], ['bar__baz.txt']),
 ('sample/test-venky/data/foo/bar', [], ['', 'bar__baz.txt'])]
```

^ notice that bar__baz.txt is returned twice, once under foo/  and once under bar/ -- the second instance is just wrong

### What did you expect to happen?

The pattern seems to be that if you have if you have a folder x/ and a file x__<whatever> as siblings, the children of x/ are hidden somehow

### lakeFS-spec version

0.11.0

### lakeFS version

0.7.1

**Comment by AdrianoKF** [comment]
Thanks for your detailed report, @indigoviolet! I'll investigate and get back to you soon.

**Comment by AdrianoKF** [comment]
@indigoviolet, I've pushed a potential fix for the issue you are seeing in PR #298.

Can you test if the fix works for you? (`pip install git+https://github.com/aai-institute/lakefs-spec.git@issue-297` should work)

The double underscores in the filename turned out to be a red herring. In a nutshell, the `ls` method was incorrectly including files that matched a prefix of its argument, which is why you were seeing the `data/foo/bar__baz.txt` file included in `ls("data/foo/bar")` since its name has `data/foo/bar` as a prefix. This error then affected all the other methods on the filesystem, which transitively rely on the results of `ls`.
